cai_causal_graph/time_series_causal_graph.py

```diff
@@ -0,0 +1,673 @@
+            # now check as described above (assume egdes are already directed)
+            edges = self.get_edges()
+            for edge in edges:
+                # first we need to extract the variable names from the nodes as the summary graph
+                # will have the variable names as nodes
+                source_node = edge.source
+                destination_node = edge.destination
+                source_variable_name = source_node.get_metadata()[self._meta_variable_name]
+                destination_variable_name = destination_node.get_metadata()[self._meta_variable_name]
+
+                assert source_variable_name is not None, 'Source variable name is None, cannot create summary graph.'
+                assert (
+                    destination_variable_name is not None
+                ), 'Destination variable name is None, cannot create summary graph.'
+
+                if source_variable_name != destination_variable_name and not summary_graph.is_edge_by_pair(
+                    (source_variable_name, destination_variable_name)
+                ):
+                    summary_graph.add_edge_by_pair((source_variable_name, destination_variable_name))
+
+            self._summary_graph = summary_graph
+
+        return self._summary_graph
+
+    def extend_graph(
+        self, backward_steps: Optional[int] = None, forward_steps: Optional[int] = None
+    ) -> TimeSeriesCausalGraph:
+        """
+        Return an extended graph.
+        Extend the graph in time by adding nodes for each variable at each time step from backward_steps to forward_steps.
+        If a backward step of n is specified, it means that the graph will be extended in order to include nodes at time -n.
+        For example, if t is a reference time, if backward_steps is 2, the graph will be extended to include nodes at time t-2.
+        This does not mean that the graph will be extended to only include nodes at time t-2, but rather that it will include
+        nodes at time t-2 and all the nodes that are connected to them as specified by the minimal graph.
+
+        If both backward_steps and forward_steps are None, return the original graph.
+
+        :param backward_steps: Number of steps to extend the graph backwards in time. If None, do not extend backwards.
+        :param forward_steps: Number of steps to extend the graph forwards in time. If None, do not extend forwards.
+        :return: Extended graph with nodes for each variable at each time step from backward_steps to forward_steps.
+        """
+        # check steps are valid (positive integers) if not None
+        assert backward_steps is None or backward_steps > 0
+        assert forward_steps is None or forward_steps > 0
+
+        # first get the minimal graph
+        minimal_graph = self.get_minimal_graph()
+
+        # create a new graph by copying the minimal graph
+        extended_graph = minimal_graph.copy()
+
+        if backward_steps is not None:
```

Comment:
maybe also worth checking that this is positive value? just a suggestion though, feel free to ignore